src/Package/Impl/DataInspect/Viewers/DebugObjectEvaluator.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Microsoft.R.Debugger;
+using Microsoft.R.Host.Client;
+using Microsoft.VisualStudio.R.Package.Repl;
+
+namespace Microsoft.VisualStudio.R.Package.DataInspect.Viewers {
+    [Export(typeof(IDebugObjectEvaluator))]
+    public sealed class DebugObjectEvaluator : IDebugObjectEvaluator {
+        private readonly IDebugSessionProvider _debugSessionProvider;
+        private readonly IRSessionProvider _rSessionProvider;
+        private DebugSession _debugSession;
+        private IRSession _rSession;
+
+        [ImportingConstructor]
+        public DebugObjectEvaluator(IRSessionProvider rSessionProvider, IDebugSessionProvider debugSessionProvider) {
+            _rSessionProvider = rSessionProvider;
+            _debugSessionProvider = debugSessionProvider;
+        }
+
+        public async Task<DebugEvaluationResult> EvaluateAsync(string expression, DebugEvaluationResultFields fields) {
+            await TaskUtilities.SwitchToBackgroundThread();
+            DebugEvaluationResult result = null;
+
+            var debugSession = await GetDebugSessionAsync();
+            var frames = await debugSession.GetStackFramesAsync();
+
+            for (int i = frames.Count - 1; i >= 0; i--) {
+                result = await frames[i].EvaluateAsync(expression, fields) as DebugValueEvaluationResult;
+                if (result != null) {
+                    break;
+                }
+            }
+            return result;
+        }
+
+        private async Task<DebugSession> GetDebugSessionAsync() {
+            if (_debugSession == null) {
+                if (_rSession == null) {
+                    _rSession = _rSessionProvider.GetInteractiveWindowRSession();
+                }
+                _debugSession = await _debugSessionProvider.GetDebugSessionAsync(_rSession);
```

Comment:
Yes, I have thought about it. `DebugSession` currently handles three unrelated things - breakpoints/stepping, stack tracing, and data/eval. The latter part doesn't really need anything from the former, and can be separated. Stack tracing is kinda related, but it is a one-way dependency (i.e. stack needs eval, but eval doesn't need stack). With my recent changes decoupling `DebugEvaluationResult` from `DebugStackFrame`, this one-way dependency is reflected in the code, as well.

The only real reason why it was all in the same place was because it used helpers from the same helper package, and so `DebugSession.InitializeAsync` was loading that package. But now the package is used everywhere else and is loaded elsewhere, so that reason is gone.

So, yeah. I have some more debugger refactoring work queued up, so timing is good too. Just please file an issue for it and assign to me so that I don't forget.